fedimint-api/src/db/mod.rs

```diff
@@ -789,6 +833,69 @@ macro_rules! push_db_key_items {
     };
 }
 
+/// MigrationMap is a BTreeMap that maps DatabaseVersions to async functions. These functions are expected
+/// to "migrate" the database from the keyed DatabaseVersion to DatabaseVersion + 1.
+pub type MigrationMap<'a> = BTreeMap<
+    DatabaseVersion,
+    for<'b> fn(
+        &'b mut DatabaseTransaction<'a>,
+    ) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send + 'b>>,
+>;
+
+/// Wrapper struct that contains the necessary data for `FedimintConsensus::new` to execute the migrations
+/// via `apply_migrations`.
+pub struct DatabaseMigrationData<'a> {
+    pub kind: ModuleKind,
+    pub target_db_version: DatabaseVersion,
+    pub migrations: MigrationMap<'a>,
+}
+
+/// `apply_migrations` iterates from the on disk database version for the module up to `target_db_version` and executes all
+/// of the migrations that exist in the `MigrationMap` of `DatabaseMigrationData`. Each migration in `MigrationMap` updates the
+/// database to have the correct on-disk structures that the code is expecting. The entire migration process is atomic
+/// (i.e migration from 0->1 and 1->2 happen atomically). This function is called before the module is initialized and as
+/// long as the correct migrations are supplied in `MigrationMap` of `DatabaseMigrationData`, the module will be able to read
+/// and write from the database successfully.
+pub async fn apply_migrations<'a>(
+    db: &'a Database,
+    migration_data: DatabaseMigrationData<'a>,
+) -> Result<(), anyhow::Error> {
+    let mut dbtx = db.begin_transaction().await;
+    let ondisk_version = dbtx.get_value(&DatabaseVersionKey).await?;
+    let target_dbversion = migration_data.target_db_version;
+    let db_version = if let Some(ondisk_version) = ondisk_version {
+        let mut current_dbversion = ondisk_version;
+
+        if current_dbversion > target_dbversion {
+            return Err(anyhow::anyhow!(format!(
+                "On disk database version for module {0} was higher than the code database version.", migration_data.kind
+            )));
+        }
+
+        while current_dbversion < target_dbversion {
+            if let Some(migration) = migration_data.migrations.get(&current_dbversion) {
+                migration(&mut dbtx).await?;
+            } else {
+                panic!("DatabaseVersion {current_dbversion} does not contain a migration.");
```

Comment:
`"Missing migration for version {}"`?

BTW. I wonder if we should index on the `to_version` or `from_version`. I guess `from` (that you currently use) have a nice property that it starts from `0`...

Reply:
This was my thinking as well. I originally had it the other way but ended up confusing myself since things appeared to be off by 1.